### PR TITLE
fix: should not change existing condition of type

### PR DESCRIPTION
**Pull request by erikgb**
When attempting to implement one of the suggested solutions to https://github.com/cert-manager/approver-policy/issues/297, I discovered that there is a bug in the code: we don't take existing certificaterequest conditions into account when preparing the SSA patch.

IMO we could implement SSA with code that is easier to read and understand. Which could reduce the chance of bugs like this one. Hopefully, we can discuss how to get some progress on https://github.com/cert-manager/trust-manager/pull/217. Since I started in trust-manager I suggest you give me a chance to show you which improvements I have in mind. If it's successful, I am happy to do a similar effort in this project.

Fixes https://github.com/cert-manager/approver-policy/issues/297 (hopefully, if not I will just reopen it)

/cc @wallrj @inteon

### CODE REVIEW

File: pkg/internal/controllers/certificaterequests.go
Comment (inteon) [inline pkg/internal/controllers/certificaterequests.go:197]: Could we instead just return without doing a patch?

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:197]: I would prefer to not do it in this PR. That will require a lot of additional changes/refactoring. 

Not a bad idea, it will only require rewriting code and test.

**Comment by erikgb** [comment]
> > Fixes [cert-manager/trust-manager#217](https://github.com/cert-manager/trust-manager/pull/217) (hopefully, if not I will just reopen it)
> 
> Is that link correct? I wouldn't expect this to fix a trust-manager PR 😁

Whooops, thanks! Fixed now.

File: pkg/internal/controllers/certificaterequests.go
Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:202]: It seems like I was a bit quick to reject your suggestion, @inteon. After a walkthrough of this PR with @wallrj in today's stand-up, I realize that your suggestion makes a lot of sense to do now, and here it is: https://github.com/cert-manager/approver-policy/pull/303/commits/bb75196b1e98dcb078eb70888795c95529cfdde6. I still think we should bugfix the conditions patch - just in case the controller cache is stale.

**Comment by erikgb** [comment]
/hold

I would like to squash/rebase this PR before merging. But please review. 😉

File: pkg/internal/controllers/certificaterequests.go
Comment (inteon) [inline pkg/internal/controllers/certificaterequests.go:203]: @erikgb do we still need these changes?

Comment (inteon) [reply pkg/internal/controllers/certificaterequests.go:203]: I think the lines 179-183 do fix the issue already right?

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:203]: Yes, either will fix issues. This is a bug IMO - regardless of the early exit.

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:203]: The early exit just hides the bug. But if the cache is stale, we will be hit if not fixing this also.

File: pkg/internal/controllers/certificaterequests.go
Comment (inteon) [inline pkg/internal/controllers/certificaterequests.go:246]: Could we use the https://github.com/cert-manager/issuer-lib/blob/main/conditions/certificaterequest.go implementation instead?

Comment (inteon) [reply pkg/internal/controllers/certificaterequests.go:246]: We fixed these issues in that repo (see eg. https://github.com/cert-manager/issuer-lib/commit/aa4880ccbbe50d10992fb103aada6402694745b2).

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:246]: I'll give it a try.

File: pkg/internal/controllers/certificaterequests.go
Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:203]: I also noticed this now "duplicated" line:

https://github.com/cert-manager/approver-policy/blob/c8f950602e17a0a6fc0d0ef78dd61db20b387d59/pkg/internal/controllers/certificaterequests.go#L100

This should probably be cleaned up. And makes me even more sure we must have the `lastTransitionTime` bugfix.

File: pkg/internal/controllers/certificaterequests.go
Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go:246]: It's a new dependency, and I am not a big fan of dependencies. 😉 But if we think this new dependency is part of the future of this project it seems to work. Incoming commit.

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go]: I stayed away from the latest version to avoid being forced to upgrade to Go 1.21.

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go]: Please review and let me know what you think. I want to rebase before we merge something. Keeping the commits until we decide how to fix the issue.

Comment (inteon) [reply pkg/internal/controllers/certificaterequests.go]: Sorry for the ambiguity, I was thinking we could copy the code from issuer-lib. I don't think we should have it as a dependency, since we are not building an issuer. What do you think?

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go]: Ok, works for me. 😄 I want to get this issue fixed. After discovering the even earlier exit, I suggest to leave that duplication out and focus on the bug here: not taking `lastTranstionTime` from and existing condition into account. I wrote a test just for that - which is a bit pointless as the PR stands now. Since we exit before it becomes relevant.

Personally, I think the code in issuer-lib is too complex and hard to read. But I really want this bug fixed.

Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go]: Copied the function from issuer-lib now.

File: pkg/internal/controllers/certificaterequests.go
Comment (erikgb) [reply pkg/internal/controllers/certificaterequests.go]: I suggest we now remove the duplicated early exit, ref.

https://github.com/cert-manager/approver-policy/blob/c8f950602e17a0a6fc0d0ef78dd61db20b387d59/pkg/internal/controllers/certificaterequests.go#L115-L118

Isn't a predicate better and sufficient?

**Review by inteon — APPROVED** [review]
Thank you for fixing this bug!
/approve
/lgtm

**Comment by erikgb** [comment]
@inteon Should I rebase before merge?

**Comment by inteon** [comment]
@erikgb sorry I thought it had been rebased already, feel free to create a new PR to fix any remaining problems.

**Pull request merged**
